File: web-admin/src/lib/youtube.ts
/**
 * We only ever store the canonical watch URL and derive everything else.
 * Thumbnails come from i.ytimg.com and titles from oEmbed, so there is no
 * YouTube Data API key to manage and no daily quota to run out of.
 */

const PATTERNS = [
  /(?:youtube\.com\/watch\?(?:.*&)?v=)([\w-]{11})/,
  /(?:youtu\.be\/)([\w-]{11})/,
  /(?:youtube\.com\/embed\/)([\w-]{11})/,
  /(?:youtube\.com\/shorts\/)([\w-]{11})/,
  /(?:youtube\.com\/live\/)([\w-]{11})/,
];

export function extractYouTubeId(url: string): string | null {
  if (!url) return null;
  for (const pattern of PATTERNS) {
    const match = url.match(pattern);
    if (match) return match[1];
  }
  return /^[\w-]{11}$/.test(url.trim()) ? url.trim() : null;
}

export function youTubeThumbnail(id: string): string {
  return `https://i.ytimg.com/vi/${id}/hqdefault.jpg`;
}

export function youTubeWatchUrl(id: string): string {
  return `https://www.youtube.com/watch?v=${id}`;
}

export interface OEmbedResult {
  title: string;
  authorName: string;
  thumbnailUrl: string;
}

export async function fetchYouTubeMeta(id: string): Promise<OEmbedResult | null> {
  try {
    const response = await fetch(
      `https://www.youtube.com/oembed?url=${encodeURIComponent(
        youTubeWatchUrl(id),
      )}&format=json`,
    );
    if (!response.ok) return null;
    const data = (await response.json()) as {
      title?: string;
      author_name?: string;
      thumbnail_url?: string;
    };
    return {
      title: data.title ?? "",
      authorName: data.author_name ?? "",
      thumbnailUrl: data.thumbnail_url ?? youTubeThumbnail(id),
    };
  } catch {
    return null;
  }
}

File: mobile_app/lib/src/core/youtube.dart
/// YouTube URL handling.
///
/// A direct port of `web-admin/src/lib/youtube.ts`, patterns included. The desk
/// stores only the canonical id and everything else is derived from it, so
/// there is no Data API key to manage and no daily quota to run out of.
///
/// Kept identical to the web version on purpose: an editor pastes a link into
/// the admin panel, and if the app's idea of a valid link differs by one
/// pattern, a video that previewed correctly renders as a grey box in the app.
abstract final class YouTube {
  const YouTube._();

  static final _patterns = <RegExp>[
    RegExp(r'(?:youtube\.com/watch\?(?:.*&)?v=)([\w-]{11})'),
    RegExp(r'(?:youtu\.be/)([\w-]{11})'),
    RegExp(r'(?:youtube\.com/embed/)([\w-]{11})'),
    RegExp(r'(?:youtube\.com/shorts/)([\w-]{11})'),
    RegExp(r'(?:youtube\.com/live/)([\w-]{11})'),
  ];

  static final _bareId = RegExp(r'^[\w-]{11}$');

  /// The 11-character id, or null if this is not a YouTube link.
  static String? extractId(String? url) {
    if (url == null || url.isEmpty) return null;
    for (final pattern in _patterns) {
      final match = pattern.firstMatch(url);
      if (match != null) return match.group(1);
    }
    final trimmed = url.trim();
    return _bareId.hasMatch(trimmed) ? trimmed : null;
  }

  static String thumbnail(String id) =>
      'https://i.ytimg.com/vi/$id/hqdefault.jpg';

  static Uri watchUri(String id) =>
      Uri.parse('https://www.youtube.com/watch?v=$id');
}
